gossip/src/crds_value.rs

```diff
@@ -490,84 +493,213 @@ impl Sanitize for NodeInstance {
+                current_bit_count = 1;
+            }
+        }
+        Self::add_number_and_test_full(&mut encoded, &mut total_serde_bytes, current_bit_count);
+        Self { encoded }
+    }
+
+    pub fn slots_count(&self) -> usize {
+        self.encoded.iter().map(|x| x.0 as usize).sum::<usize>()
+    }
+
+    pub fn to_slots(&self, last_slot: Slot, min_slot: Slot) -> Vec<Slot> {
+        let mut result: Vec<Slot> = Vec::new();
+        let mut offset = 0;
+        let mut current_bit = 1;
+        for bit_count in &self.encoded {
+            let count: u16 = bit_count.0;
+            if current_bit > 0 {
+                for i in 0..count {
+                    let slot = last_slot.saturating_sub(offset).saturating_sub(i as Slot);
+                    if slot < min_slot {
+                        return result;
+                    }
+                    result.push(slot);
+                    if result.len() > MAX_SLOTS_PER_ENTRY {
+                        break;
+                    }
+                }
+            }
+            offset += count as Slot;
+            current_bit = 1 - current_bit;
+        }
+        result.reverse();
+        result
+    }
+}
+
+#[derive(Deserialize, Serialize, Clone, Debug, PartialEq, Eq, AbiExample)]
+pub struct RawOffsets {
+    bits: BitVec<u8>,
+}
+
+impl RawOffsets {
+    pub fn new(mut bits: BitVec<u8>) -> Self {
+        bits.truncate(RestartLastVotedForkSlots::MAX_SPACE as u64 * 8);
+        bits.shrink_to_fit();
+        Self { bits }
+    }
+
+    pub fn to_slots(&self, last_slot: Slot, min_slot: Slot) -> Vec<Slot> {
+        let mut result = Vec::new();
+        for i in 0..self.bits.len() {
+            if self.bits.get(i) {
+                let slot = last_slot.saturating_sub(i as Slot);
+                if slot < min_slot {
+                    return result;
+                }
+                result.push(slot);
+                if result.len() > MAX_SLOTS_PER_ENTRY {
+                    break;
+                }
+            }
+        }
+        result.reverse();
+        result
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone, PartialEq, Eq, AbiExample, Debug)]
 pub struct RestartLastVotedForkSlots {
     pub from: Pubkey,
     pub wallclock: u64,
-    pub slots: Vec<CompressedSlots>,
+    offsets: SlotsOffsets,
+    pub last_voted_slot: Slot,
     pub last_voted_hash: Hash,
     pub shred_version: u16,
 }
 
 impl Sanitize for RestartLastVotedForkSlots {
     fn sanitize(&self) -> std::result::Result<(), SanitizeError> {
-        if self.slots.is_empty() {
-            return Err(SanitizeError::InvalidValue);
-        }
-        self.slots.sanitize()?;
         self.last_voted_hash.sanitize()
     }
 }
 
 impl RestartLastVotedForkSlots {
-    pub fn new(from: Pubkey, now: u64, last_voted_hash: Hash, shred_version: u16) -> Self {
-        Self {
+    // This number is MAX_CRDS_OBJECT_SIZE - empty serialized RestartLastVotedForkSlots.
+    const MAX_SPACE: usize = 824;
```

Comment:
You may be able to define this in terms of the other constants and use `const_assert_eq!()` to make this less fragile. There are some similar patterns in `legacy.rs`.

Reply:
Okay, I'll take a look.